server/api/impl/admin/synchronize.go

```diff
@@ -0,0 +1,225 @@
+		manifest, err := adm.store.LoadManifest(registeredApp.Manifest.AppID)
+		if err != nil {
+			return errors.Wrapf(err, "can't load manifest from store appID = %s", registeredApp.Manifest.AppID)
+		}
+		if err := adm.DeleteApp(registeredApp); err != nil {
+			return errors.Wrapf(err, "can't delete an app")
+		}
+		if manifest.Version != registeredApp.Manifest.Version {
+			// update app in proxy plugin
+			oldVersion := registeredApp.Manifest.Version
+			registeredApp.Manifest = manifest
+			if err := adm.store.StoreApp(registeredApp); err != nil {
+				return errors.Wrapf(err, "can't store app - %s", registeredApp.Manifest.AppID)
+			}
+			updatedAppVersionsMap[registeredApp.Manifest.AppID] = oldVersion
+		}
+	}
+
+	// Add new apps as registered
+	for appID, manifest := range adm.store.ListManifests() {
+		if _, ok := registeredAppsMap[appID]; ok {
+			continue
+		}
+		if err := adm.AddApp(manifest); err != nil {
+			return errors.Wrapf(err, "can't add new app as registered appID - %s", manifest.AppID)
+		}
+	}
+
+	registeredAppsUpgraded := adm.store.ListApps()
+
+	// call onInstanceStartup. App migration happens here
+	for _, registeredApp := range registeredAppsUpgraded {
+		if registeredApp.Status == apps.AppStatusEnabled {
+			values := map[string]string{}
+			if _, ok := updatedAppVersionsMap[registeredApp.Manifest.AppID]; ok {
+				values[oldVersionKey] = updatedAppVersionsMap[registeredApp.Manifest.AppID]
+			}
+			// Call onStartup the function of the app. It should be called only once
+			f := func() error {
+				if err := adm.call(registeredApp, registeredApp.Manifest.OnStartup, values); err != nil {
+					adm.mm.Log.Error("Can't call onStartup func of the app", "app_id", registeredApp.Manifest.AppID, "err", err.Error())
+				}
+				return nil
+			}
+			if err := adm.callOnce(f); err != nil {
+				adm.mm.Log.Error("Can't callOnce the onStartup func of the app", "app_id", registeredApp.Manifest.AppID, "err", err.Error())
+			}
+		}
+	}
+
+	return nil
+}
+
+func (adm *Admin) DeleteApp(app *apps.App) error {
+	// Call delete the function of the app
+	if err := adm.call(app, app.Manifest.OnDelete, nil); err != nil {
+		return errors.Wrapf(err, "delete failed. appID - %s", app.Manifest.AppID)
+	}
+
+	// delete oauth app
+	conf := adm.conf.GetConfig()
+	client := model.NewAPIv4Client(conf.MattermostSiteURL)
+	client.SetToken(string(adm.adminToken))
+
+	if app.OAuth2ClientID != "" {
+		success, response := client.DeleteOAuthApp(app.OAuth2ClientID)
+		if !success || response.StatusCode != http.StatusNoContent {
+			return errors.Wrapf(response.Error, "failed to delete OAuth2 App - %s", app.Manifest.AppID)
+		}
+	}
+
+	// delete app from proxy plugin, not removing the data
+	if err := adm.store.DeleteApp(app); err != nil {
+		return errors.Wrapf(err, "can't delete app - %s", app.Manifest.AppID)
+	}
+
+	adm.mm.Log.Info("Deleted the app", "app_id", app.Manifest.AppID)
+
+	return nil
+}
+
+func (adm *Admin) AddApp(manifest *apps.Manifest) error {
```

Comment:
Maybe instead have `ListApp' and rename the previous `LoadApp` as `GetListedApp`